edx/edx-enterprise
# ENT-3227 : Client Impl + Adding oauth access token retrieval

**Merge checklist:**
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.

**Post merge:**
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

**What this PR does:**
- [x] Custom field `refresh_token` added to the CanvasEnterpriseCustomerConfiguration model
- [x] Field `oauth_api_path ` moved from model to app config
- [x] Client now has the ability to create a session 
      - [ ] Sessions, upon instantiation, retrieve and save the access token associated with the enterprise user.

**What work is left after this merges:**
- Related Issue [ENT-3303](https://openedx.atlassian.net/browse/ENT-3303)
- [ ] Add another endpoint that will be hit when the user authorizes us to access their Canvas account, to do this, they will need to add this new endpoint as the redirect path on their developer key. 
     - I see the flow for a user setting up the necessary credentials for us to integrate Canvas as such:
              - User's generate a Developer Key through Canvas that contains both a client ID and secret. 
              - User adds our (TODO) `/oauth_complete` endpoint in the setting's of their developer key `redirect_uri`
              - The user needs to hit the endpoint `<user's canvas_base_url>/login/oauth2/auth?client_id=<user's client ID>&response_type=code&redirect_uri=<(TODO) our /oauth_complete endpoint>`
                        - Right now, I can see this happening a couple of ways. We could either have ECS provide them with the link and ask them to just copy and paste it into their browser, or we could possibly build out a UI in the admin frontend portal where they provide the client ID and hit a button labeled `Generate Oauth code`. Also very open to other suggestions here. 


**Steps to test/use locally:**
- fetch and pull this branch
- run migrations so that you have the the new field `refresh_token` in the canvas model
- Have a local instance of Canvas running (https://github.com/instructure/canvas-lms/wiki) 
- Create a Canvas admin user with developer key creation capabilities
- Generate a developer key
   - go into the settings of the dev key and add a redirect uri (before the oauth/complete endpoint is complete, it doesn't matter what this uri is, as long as you remember it.
- using the secret and client id from the developer key visit:
   - <your local canvas base path>/login/oauth2/auth?client_id=<your client id>&response_type=code&redirect_uri=<your redirect uri>
- when prompted, confirm authorization
- after authorizing, you will be redirected to whatever redirect uri was provided
   - Save the code passed in at the end of the redirect uri, you will use this to get the access token and refresh token
- Make a post request to `http://localhost:3000/login/oauth2/token` with the following params
   -  `grant_type` : `authorization_code`
   -  `client_id` : < your client id >
   -  `client_secret` : < your client secret >
   -  `redirect_uri` : < the same redirect uri that you saved to your developer key >
   -  `code` : < the code you got after authorizing >
- Copy both the access token and the refresh token 
   - the access token is your auth token that will allow you to post/delete courses to Canvas
- Create a new Enterprise user in edx-enterprise and create a new CanvasEnterpriseCustomerConfiguration and CanvasGlobalConfiguration model for the user, using the client id, secret, refresh token, base url, account id, course api path
- you should now be able to generate a new client session that retrieves and stores the access token in order to post and delete to your canvas instance.
